### PR TITLE
Switch to proto-lens

### PR DESCRIPTION
**Pull request by mitchellwrosen**
#118 

This patch just touches `riak-protobuf`, not `riak`.

/cc @llelf Thoughts on this change?

### CODE REVIEW

**Comment by llelf** [comment]
👍 
but why one `riak.proto` (in riak2 too btw)?

**Comment by mitchellwrosen** [comment]
Oh, because for a bit I was tweaking the message/field names to my taste, but then decided that naming things is too hard and it would be better to use exactly what Basho defined. I will swap riak.proto with the individual files, the generated code will be exactly the same.

**Comment by mitchellwrosen** [comment]
Hm, multiple files cause multiple `_Fields.hs` output files that define the same lens, e.g.

```
bucket ::
       forall f s a .
         (Prelude.Functor f, Lens.Labels.HasLens' s "bucket" a) =>
         Lens.Family2.LensLike' f s a
bucket
  = Lens.Labels.lensOf'
      ((Lens.Labels.proxy#) :: (Lens.Labels.Proxy#) "bucket")
```

**Comment by mitchellwrosen** [comment]
CI won't be happy until I fix `riak` :)

**Comment by llelf** [comment]
@mitchellwrosen  how about `echo 'cat *.proto > all-riak.proto' >> generate.sh` and then just using one file?

**Comment by mitchellwrosen** [comment]
@llelf Sure

File: protobuf/src-proto/Proto/Proto/Riak.hs [hunk 1]
```diff
@@ -0,0 +1,11938 @@
+{- This file was auto-generated from proto/riak.proto by the proto-lens-protoc program. -}
+{-# LANGUAGE ScopedTypeVariables, DataKinds, TypeFamilies,
+  UndecidableInstances, GeneralizedNewtypeDeriving,
+  MultiParamTypeClasses, FlexibleContexts, FlexibleInstances,
+  PatternSynonyms, MagicHash, NoImplicitPrelude, DataKinds #-}
+{-# OPTIONS_GHC -fno-warn-unused-imports#-}
+{-# OPTIONS_GHC -fno-warn-duplicate-exports#-}
```
Comment (tmcgilchrist) [inline protobuf/src-proto/Proto/Proto/Riak.hs]: If these `OPTIONS_GHC` are included here we can drop them from the `ghc-options:` section of the cabal file.

File: protobuf/riak-protobuf.cabal [hunk 2]
```diff
@@ -15,122 +15,29 @@ copyright:           Copyright 2011 MailRank, Inc.
...
-  proto.sh
-  src/riak.proto
-  src/riak_kv.proto
-  src/riak_search.proto
-  src/riak_dt.proto
-  src/riakextra.proto
   Changes.md
+  README.markdown
+  generate-proto.sh
+  proto/riak.proto
```
Comment (tmcgilchrist) [inline protobuf/riak-protobuf.cabal]: How much do we value matching the file names from the *official* source? https://github.com/bet365/riak_pb/tree/develop/src

Not that they seem to have changed in the last 3+ years.

Comment (mitchellwrosen) [reply protobuf/riak-protobuf.cabal]: Per @llelf I will be reverting this to use multiple files

**Comment by mitchellwrosen** [comment]
Got `riak` compiling, still have to fix its test suite.

**Comment by mitchellwrosen** [comment]
Any idea what to do about this one? I'm not sure what's pulling in `entwine`.

```
src/Entwine/P.hs:253:18:
    Could not find module ‘Data.Semigroup’
    It is a member of the hidden package ‘semigroups-0.18.5@JoK7OkoWGDAB27bANwddFj’.
    Perhaps you need to add ‘semigroups’ to the build-depends in your .cabal file.
    Use -v to see a list of the files searched for.
```

**Comment by mitchellwrosen** [comment]
I was able to build successfully on GHCs going back to 8.0 manually, but it isn't for me to test on cabal 1.24 + ghc 7.10, so relying on CI here. But unrelated to the semigroup error are build errors due to `-Werror`, which I will fix soon.

**Comment by tmcgilchrist** [comment]
That’s an error related to ghc 7.10 support. Out and about right now so
can’t included the details.

I propose we drop 7.10 support in this new release. What do you think?

**Comment by mitchellwrosen** [comment]
@tmcgilchrist Agreed, supporting 3-4 major GHC versions is plenty.

**Comment by mitchellwrosen** [comment]
Whoops, looks like I broke CI by updating `.travis.yml` to use `cabal-install-2.4`.

```
Warning: The sandbox command is a part of the legacy v1 style of cabal usage.
Please switch to using either the new project style and the new-sandbox
command or the legacy v1-sandbox alias as new-style projects will become the
default in the next version of cabal-install. Please file a bug if you cannot
replicate a working v1- use case with the new-style commands.
```

**Comment by mitchellwrosen** [comment]
Could someone familiar with `mafia` chime in? Am I supposed to build with a sufficiently old ver of cabal, like 2.0?

**Comment by llelf** [comment]
I used to use mafia.
@tmcgilchrist does mafia needed at all right now? cabal new-build does everything better

**Comment by tmcgilchrist** [comment]
Probably easiest to just switch to a ci based off
https://github.com/haskell-CI/haskell-ci

There are some features that you miss on the dev side like sub project
reloading and extra warning but I’m probably the only one using those. :-)

**Comment by tmcgilchrist** [comment]
FYI @mitchellwrosen There's an open issue with supporting Cabal 2.2 / 2.4 in mafia https://github.com/haskell-mafia/mafia/pull/234

Cabal new-build has grown some of the capabilities like nix style sandboxing / dependency sharing that we built into mafia. Ideally we would swap out the piece we can in mafia using the cabal version where possible and just maintain the pieces that need to be different. @llelf

**Comment by tmcgilchrist** [comment]
https://github.com/hedgehogqa/haskell-hedgehog/blob/master/.travis.yml from `hedgehog` is a good example of how to make it work.

Looks like you're shaving the whole yak in one session with this PR :-D

**Comment by mitchellwrosen** [comment]
Whoops, I messed with the script too much and now the `tools` submodule isn't there in the `before_script` phase... but at a glance I can't quite tell why...

```
$ sudo ./tools/travis-ci/riak-install -d "$RIAK_DOWNLOAD_URL"
sudo: ./tools/travis-ci/riak-install: command not found
```

**Comment by mitchellwrosen** [comment]
Before (good):

[.travis.yml](https://github.com/riak-haskell-client/riak-haskell-client/blob/e22e3fce5c5c4e769840b7413d5cc163dc130d77/.travis.yml) 
[result](https://travis-ci.org/riak-haskell-client/riak-haskell-client/jobs/445994283)

After (bad):

[.travis.yml](https://github.com/riak-haskell-client/riak-haskell-client/blob/8191b7e53110f39eb5f7a87d6f951a48249cf97c/.travis.yml)
[result](https://travis-ci.org/riak-haskell-client/riak-haskell-client/jobs/497305005)

**Comment by llelf** [comment]
@mitchellwrosen 
wrong permissions for `./tools/travis-ci/riak-install`
or do  `sudo sh ./tools/travis-ci/riak-install`

File: .travis.yml [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
+#   haskell-ci 'travis' 'cabal.project'
+#
+# For more information, see https://github.com/haskell-CI/haskell-ci
+#
 language: c
-sudo: false
+dist: xenial
 
 git:
   submodules: false  # whether to recursively clone submodules
```
Comment (tmcgilchrist) [inline .travis.yml]: You might need this as `true` to get riak `tools`?

Comment (mitchellwrosen) [reply .travis.yml]: @tmcgilchrist Worth a shot but it was `false` before :)

@llelf's suggestion of `sudo sh` did not work, trying again with `sudo: false` this time (as before), I'm thinking it is going to to break again, though

**Comment by llelf** [comment]
@mitchellwrosen `sudo: false` is absolutely necessary; travis will provide more memory that way

File: .travis.yml (hunk 3 above)
Comment (mitchellwrosen) [reply .travis.yml]: This was the problem after all. I guess mafia was taking care of cloning submodules previously.

Comment (tmcgilchrist) [reply .travis.yml]: yeah `mafia` does that for you.

**Comment by mitchellwrosen** [comment]
Cool, CI is compiling, one last thing for me to do - split up `riak.proto`.

**Comment by mitchellwrosen** [comment]
I believe this is ready to go, just that one `network` bounds concern. The released bounds of `1.1.2.6` are innacurate; it says `network >= 2.3` but actually requires `network >= 3.0`. So my options for this patch are:
- Revert to `>= 2.3 && < 3.0`
- Use CPP to allow `>= 2.3`
- Require `>= 3.0` (which is what this patch currently contains)

**Comment by mitchellwrosen** [comment]
@tmcgilchrist Ping, just that one issue from my previous comment remains.

**Comment by mitchellwrosen** [comment]
I'm just going to merge, if someone complains about the `network >= 3.0` lower bound they can provide a patch for `< 3.0` compat :)

**Comment by tmcgilchrist** [comment]
Nice work 👍

**Pull request merged**